### PR TITLE
FIX upgrade Flask to 1.0.2 in ci dockerfile

### PR DESCRIPTION
**Pull request by fgalan**
We have recently found a bug in default Flask version installed by default in CentOS7 which impact in the NGSIv2 query forwarding functionality (Flask is being used by the accumulator-server.py script using in functional tests). This PR upgrades Flask version to 1.0.2, which doesnt' have the bug, using virtualenv.

@caa06d9c could you have a look and provide feedback and/or LGTM, pls? Given this PR touches the ci stuff you authored, I think you are the most appropiate reviewer :)

### CODE REVIEW

File: ci/rpm7/build-dep.sh [hunk 1]
```diff
@@ -78,6 +78,17 @@ curl -L https://nexus.lab.fiware.org/repository/raw/public/storage/gmock-1.5.0.t
 #  && make install \
 #  && rm -Rf /opt/mosquitto-1.5
 
+# CentOS 7 installs Flask==0.10.1 which depends on Werkzeug==0.9.1. There is a bug
+# in Werkzeug which makes an empty content-length header to appear in the accumulator-server.py
+# dumps. The bug is fixed in Werkzeug==0.11.16. Thus, we override the system setting,
+# installing in the virtual env Flask==1.0.2, which depends on Werkzeug==0.15.2
+cd /opt \
```
Comment (fgalan) [inline ci/rpm7/build-dep.sh]: At first glance this `cd` could seem useless. However, if not included pip install fails complaning about the directory from which it is being executed doesn't exit (I don't remember the exact wording).

NTC (just informative)

**Comment by fgalan** [comment]
With regards to travis fail, it seems we have a "chicken and egg" problem here. Travis reports:

```
Builder: build functional
------------------------------------- prepare  -----------------------------------------
./scripts/build/compileInfo.sh --release
diff: src/lib/common/compileInfo.h: No such file or directory
. scripts/testEnv.sh
. /opt/ft_env/bin/activate
/bin/sh: /opt/ft_env/bin/activate: No such file or directory
make: *** [prepare] Error 1
Builder: make failed
The command "if [ "$TEST" != "compliance" ]; then docker run -t --rm -v $(pwd):/opt/fiware-orion $IMAGE build -ms $TEST; fi" failed and exited with 1 during .
```

which makes sense, as the image at https://hub.docker.com/r/fiware/orion-ci has not been build yet based  in the modifications in this PR.

Thus, I guess travis should be ignored in this PR (or manually build fiware/orion-ci:rpm7 based in the content of the branch PR, then relaunch travis).

By the way, looking https://hub.docker.com/r/fiware/orion-ci/tags it seems :rpm7 is 7 months olds. Is this intentionally? (I have the "filling" that is should be configured to automatically re-build each new commit in master)

**Comment by caa06d9c** [comment]
LGTM

File: test/functionalTest/cases/2015_notification_templates/notification_templates_cache_refresh.test [hunk 2]
```diff
@@ -130,13 +130,13 @@ Date: REGEX(.*)
 ============================================
 PATCH http://127.0.0.1:REGEX(\d+)/notify?a1=a1&a3=13&a2=a2&type=Thing&id=E1
 Fiware-Servicepath: /
-Ngsiv2-Attrsformat: custom
```
Comment (fgalan) [inline test/functionalTest/cases/2015_notification_templates/notification_templates_cache_refresh.test]: Based on last travis check, the changes in this and the other .test file are needed to make ft to work. Some headers change their order.

This is a bit weird and I'd like to confirm is ok. However, it goes beyond my knowledgue, as it seems related with some internal in Flask or Werkzeug). I have asked in SOF https://stackoverflow.com/questions/55734118/flask-werkzeug-change-headers-order-after-upgrade. Please upvote ;)

Comment (fgalan) [reply test/functionalTest/cases/2015_notification_templates/notification_templates_cache_refresh.test]: Based in the "answer" of that question and this one https://stackoverflow.com/questions/55751419/getting-request-headers-in-the-same-order-they-are-received-in-flask/55752367#55752367, is fine for me *in the context of this PR*.

*For the future* I have created an issue to think about it: https://github.com/telefonicaid/fiware-orion/issues/3473

Comment (kzangeli) [reply test/functionalTest/cases/2015_notification_templates/notification_templates_cache_refresh.test]: Any idea why these two tests got changes in HTTP headers order and no other tests?

File: test/functionalTest/README.md [hunk 3]
```diff
@@ -1,6 +1,55 @@
...
+
+```
+pip install Flask==1.0.2
+pip install pyOpenSSL==19.0.0
+```
+
+Next, install the accumulator-server.py script itself:
+
+```
+make install_scripts  # add INSTALL_DIR=... if you need to install in an specific place
```
Comment (kzangeli) [inline test/functionalTest/README.md]: Typo: an -> a

File: ci/rpm7/build-dep.sh [hunk 4]
```diff
@@ -78,6 +77,18 @@ curl -L https://nexus.lab.fiware.org/repository/raw/public/storage/gmock-1.5.0.t
 #  && make install \
 #  && rm -Rf /opt/mosquitto-1.5
 
+# CentOS 7 installs Flask==0.10.1 which depends on Werkzeug==0.9.1. There is a bug
+# in Werkzeug which makes an empty content-length header to appear in the accumulator-server.py
```
Comment (kzangeli) [inline ci/rpm7/build-dep.sh]: to appear => appear

File: ci/rpm7/makefile [hunk 5]
```diff
@@ -87,7 +87,10 @@ install_functional: install
 functional:
 	@echo '------------------------------------- make functional ----------------------------------'
 
-	./test/functionalTest/testHarness.sh --stopOnError
+	# ft_env needed for accumulator-server.py. Very import to "chain" the two commands with '; \'
```
Comment (kzangeli) [inline ci/rpm7/makefile]: import => important ?

File: scripts/accumulator-server.py [hunk 6]
```diff
@@ -396,13 +401,16 @@ def getPid():
...
-      # fail in some configurations (the current one is an attempt to make this to work at jenkins)
-      context = SSL.Context(SSL.SSLv23_METHOD)
-      context.use_privatekey_file(key_file)
-      context.use_certificate_file(cert_file)
-      #context = (cert_file, key_file)
+      # Commented lines correspond to the way of using context with pyOpenSSL 0.13.1 (the one that comes with system Python in CentOS 7).
+      # We are using the new way (which is simpler) as we moved to pyOpenSSL 19.0.0. Referecence: Reference: http://stackoverflow.com/questions/28579142/attributeerror-context-object-has-no-attribute-wrap-socket/28590266
+      #
+      # We need to upgrade pyOpenSSL version due to problems of installing 0.13.1 inside virtualenv. Installing the module
+      # requires to compile some parts and this cause a conflict with base openssl devel libraries in CentOS 7 (solvable by hack, but we want to avoid it)
```
Comment (kzangeli) [inline scripts/accumulator-server.py]: cause -> causes

File: test/functionalTest/README.md [hunk 7]
```diff
@@ -1,6 +1,55 @@
...
+* The `contextBroker` binary is in execution path
+* MongoDB database is up and running
+* accumulator-server.py is correctly installed and available in the execution path (see specific section below)
+
+### How to install accumulator-server.py
+
+We recommend to use Python [virtualenv](https://virtualenv.pypa.io/en/latest) to install the required dependencies to avoid any potential conflict with operating system Python installation. So, first create your virtual env (named for instance `ft_env`):
+
+```
+pip install virtualenv  # if you doesn't have virtualenv itself previously installed
```
Comment (kzangeli) [inline test/functionalTest/README.md]: doesn't -> don't
itself -> {} (just remove the word)

File: test/functionalTest/README.md [hunk 8]
```diff
@@ -1,6 +1,55 @@
...
+**IMPORTANT:** remember to activate the virtual env (`. /path/to/ft_env/bin/activate`) before running functional tests
+
+### Run functional tests
+
+The easiest way is running just:
+
+```
+cd test/functionalTest
+./testHarness.sh
+```
```
Comment (kzangeli) [inline test/functionalTest/README.md]: A thought ...
Perhaps the script testHarness.sh should create the virtualenv and start it ...  ?

File: test/functionalTest/cases/2015_notification_templates/notification_templates_cache_refresh.test (hunk 2 above)
Comment (fgalan) [reply test/functionalTest/cases/2015_notification_templates/notification_templates_cache_refresh.test]: Yes. In the SOF answers and links to duplicated questions.

File: test/functionalTest/README.md (hunk 7 above)
Comment (fgalan) [reply test/functionalTest/README.md]: Ok, the first, no the second as I want to emphatise on virtualenv. I mean, "si necesitas instalar el propio virtualenv previamente" vs "si necesitas instalar virtualenv previamente".

File: test/functionalTest/README.md (hunk 8 above)
Comment (fgalan) [reply test/functionalTest/README.md]: Its better to have testHarness.sh simple and focused in its task (run the test), leaving environment preparation aspects to others.

NTC

File: test/functionalTest/README.md (hunk 3 above)
Comment (fgalan) [reply test/functionalTest/README.md]: Fixed in 68b58b8

File: ci/rpm7/build-dep.sh (hunk 4 above)
Comment (fgalan) [reply ci/rpm7/build-dep.sh]: Fixed in 68b58b8

File: ci/rpm7/makefile (hunk 5 above)
Comment (fgalan) [reply ci/rpm7/makefile]: Fixed in 68b58b8

File: scripts/accumulator-server.py (hunk 6 above)
Comment (fgalan) [reply scripts/accumulator-server.py]: Fixed in 68b58b8

File: test/functionalTest/README.md (hunk 7 above)
Comment (fgalan) [reply test/functionalTest/README.md]: Fixed in 68b58b8

**Comment by kzangeli** [comment]
LGTM

**Comment by caa06d9c** [comment]
LGTM

File: test/functionalTest/cases/2015_notification_templates/notification_templates_cache_refresh.test (hunk 2 above)
Comment (fgalan) [reply test/functionalTest/cases/2015_notification_templates/notification_templates_cache_refresh.test]: NTC

**Comment by fgalan** [comment]
CC: @fisuda (as this PR touches walkthrough_apiv2.md so I understand that sync in the Japanese translation should be done, please)

**Comment by fisuda** [comment]
@fgalan ,
Noted with thanks.

**Pull request merged**